common/src/main/java/io/netty/util/Recycler.java

```diff
@@ -189,558 +167,104 @@ public final boolean recycle(T o, Handle<T> handle) {
-            this.maxCapacity = maxCapacity;
-            availableSharedCapacity = new AtomicInteger(max(maxCapacity / maxSharedCapacityFactor, LINK_CAPACITY));
-            elements = new DefaultHandle[min(INITIAL_CAPACITY, maxCapacity)];
-            this.interval = interval;
-            this.delayedQueueInterval = delayedQueueInterval;
-            handleRecycleCount = interval; // Start at interval so the first one will be recycled.
-            this.maxDelayedQueues = maxDelayedQueues;
-        }
-
-        // Marked as synchronized to ensure this is serialized.
-        synchronized void setHead(WeakOrderQueue queue) {
-            queue.setNext(head);
-            head = queue;
-        }
-
-        int increaseCapacity(int expectedCapacity) {
-            int newCapacity = elements.length;
-            int maxCapacity = this.maxCapacity;
-            do {
-                newCapacity <<= 1;
-            } while (newCapacity < expectedCapacity && newCapacity < maxCapacity);
-
-            newCapacity = min(newCapacity, maxCapacity);
-            if (newCapacity != elements.length) {
-                elements = Arrays.copyOf(elements, newCapacity);
-            }
-
-            return newCapacity;
-        }
+        private final int ratioInterval;
+        private final Queue<DefaultHandle<T>> pooledHandles;
+        private int ratioCounter;
 
-        @SuppressWarnings({ "unchecked", "rawtypes" })
-        DefaultHandle<T> pop() {
-            int size = this.size;
-            if (size == 0) {
-                if (!scavenge()) {
-                    return null;
-                }
-                size = this.size;
-                if (size <= 0) {
-                    // double check, avoid races
-                    return null;
-                }
-            }
-            size --;
-            DefaultHandle ret = elements[size];
-            elements[size] = null;
-            // As we already set the element[size] to null we also need to store the updated size before we do
-            // any validation. Otherwise we may see a null value when later try to pop again without a new element
-            // added before.
-            this.size = size;
-
-            if (ret.lastRecycledId != ret.recycleId) {
-                throw new IllegalStateException("recycled multiple times");
-            }
-            ret.recycleId = 0;
-            ret.lastRecycledId = 0;
-            return ret;
-        }
-
-        private boolean scavenge() {
-            // continue an existing scavenge, if any
-            if (scavengeSome()) {
-                return true;
-            }
-
-            // reset our scavenge cursor
-            prev = null;
-            cursor = head;
-            return false;
+        LocalPool(int maxCapacity, int ratioInterval) {
+            this.maxCapacity = maxCapacity;
+            this.ratioInterval = ratioInterval;
+            pooledHandles = PlatformDependent.newMpscQueue();
```

Comment:
I would expose chunk size of the queue: it impacts the footprint in memory while the pool is in steady state ie contains < next power of 2 chunk size  elements

https://github.com/JCTools/JCTools/blob/master/jctools-core/src/main/java/org/jctools/queues/MpscGrowableArrayQueue.java#L28 is a good fit as well: it would left the chunk capacity to change accordly to the load (max reached capacity) until it would be stable; but meaning that its footprint will be the max capacity ever reached.